x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/upgrade_review_algorithms.md

```diff
@@ -225,11 +240,54 @@ Examples:
 | array of scalars | index      | ["logs-*"]                            | ["logs-*", "Logs-*"]    | ["logs-*", "new-*"]              | ["logs-*", "Logs-*", "new-*"]    |
 ```
 
+#### **Scenario: `ABC` - Rule field is a solvable `data_source` object**
+
+**Automation**: 2 integration tests with mock rules + a set of unit tests for the algorithm
+
+```Gherkin
+Given data_source field is customized by the user (current version != base version)
+And data_source field is updated by Elastic in this upgrade (target version != base version)
+And customized data_source field is different than the Elastic update in this upgrade (current version != target version)
+And current version and target version are both array fields in data_source
+Then for data_source field the diff algorithm should output a custom merged version with a solvable conflict
+And arrays should be deduplicated before comparison
+And arrays should be compared sensitive of case
+And arrays should be compared agnostic of order
+And data_source field should be returned from the `upgrade/_review` API endpoint
+And data_source field should be shown in the upgrade preview UI
+
+Examples:
+| algorithm    | base_version                                                        | current_version                                                            | target_version                                                     | merged_version                                                              |
+| data_source  | {type: "index_patterns", "index_patterns": ["one", "two", "three"]} | {type: "index_patterns", "index_patterns": ["two", "one", "four"]}         | {type: "index_patterns", "index_patterns": ["one", "two", "five"]} | {type: "index_patterns", "index_patterns": ["one", "two", "four", "five"]}  |
+| data_source  | {type: "data_view", "data_view_id": "A"}                            | {type: "index_patterns", "index_patterns": ["one", "one", "two", "three"]} | {type: "index_patterns", "index_patterns": ["one", "two", "five"]} | {type: "index_patterns", "index_patterns": ["one", "two", "three", "five"]} |
```

Comment:
Is this extra `"one"` in `current_version` intentional?

Reply:
Yeah, it's similar to the scalar array diff algorithm where we test deduplication, etc., in the arrays. The tests aren't as exhaustive for this algorithm since we're reusing a lot of logic for array comparison and merging but it's still a good idea to test a little bit of it